src/plugins/data/common/data_frames/utils.ts

```diff
@@ -472,3 +472,77 @@ export const dataFrameToSpec = (dataFrame: IDataFrame, id?: string): IndexPatter
     fields: fields.reduce(flattenFields, {} as IndexPatternFieldMap),
   };
 };
+
+type FetchFunction<T, P = void> = (params?: P) => Promise<T>;
+
+export interface PollingConfigurations {
+  tabId: string;
+}
+
+export class UsePolling<T, P = void> {
+  public data: T | null = null;
+  public error: Error | null = null;
+  public loading: boolean = true;
+  private shouldPoll: boolean = false;
+  private intervalRef?: NodeJS.Timeout;
+
+  constructor(
+    private fetchFunction: FetchFunction<T, P>,
+    private interval: number = 5000,
+    private onPollingSuccess?: (data: T, configurations: PollingConfigurations) => boolean,
+    private onPollingError?: (error: Error) => boolean,
+    private configurations?: PollingConfigurations
+  ) {}
+
+  async fetchData(params?: P) {
+    this.loading = true;
+    try {
+      const result = await this.fetchFunction(params);
+      this.data = result;
+      this.loading = false;
+
+      if (this.onPollingSuccess && this.onPollingSuccess(result, this.configurations!)) {
+        this.stopPolling();
+      }
+    } catch (err) {
+      this.error = err as Error;
+      this.loading = false;
+
+      if (this.onPollingError && this.onPollingError(this.error)) {
+        this.stopPolling();
+      }
+    }
+  }
+
+  async startPolling(params?: P) {
+    this.shouldPoll = true;
+    if (!this.intervalRef) {
+      this.intervalRef = setInterval(() => {
+        if (this.shouldPoll) {
+          this.fetchData(params);
+        }
+      }, this.interval);
+    }
+  }
+
+  stopPolling() {
+    this.shouldPoll = false;
+    if (this.intervalRef) {
+      clearInterval(this.intervalRef);
+      this.intervalRef = undefined;
+    }
+  }
+
+  async waitForPolling(): Promise<void> {
+    return new Promise<void>((resolve) => {
+      const checkLoading = () => {
+        if (!this.loading) {
+          resolve();
+        } else {
+          setTimeout(checkLoading, 5000);
```

Comment:
is this the timeout logic? can we get an advanced setting for this as well? but could you explain if when we are timing out? if it is loading and then 5 seconds?

Reply:
This is the logic I used to wait for the polling response, was having issues before where the UI wouldn't display the data since it was polling. I can add a timeout though to cancel long queries